File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        def convert1Dto2D(val, rows, cols):
            r = val // cols
            c = val % cols
            return(r, c)

        l, r = 0, (len(matrix) * len(matrix[0])) -1

        while l <= r:
            mid = l + ((r-l) // 2)

            row, col = convert1Dto2D(mid, len(matrix), len(matrix[0]))

            if matrix[row][col] < target: l = mid + 1

            elif matrix[row][col] > target: r = mid - 1

            else: return True

        return False

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp

class Solution {
public:
    const bool searchMatrix(const vector<vector<int>>& matrix, const int target) {
        ROWS = matrix.size();
        COLS = matrix[0].size();
        int l = 0;
        int r = (ROWS * COLS) - 1;
        while (l <= r) {
            const int mid = l + ((r - l) / 2);
            const pair<int, int> midCoord = convert1DTo2D(mid);
            const int value = matrix[midCoord.first][midCoord.second];
            if (value > target) {
                r = mid - 1;
            } else if (value < target) {
                l = mid + 1;
            } else {
                return true;
            }
        }
        return false;
    }
private:
    const pair<int, int> convert1DTo2D(int index) {
        if (index == 0) return pair<int, int>{0, 0};
        return pair<int, int>{index / COLS, index % COLS};
    }
    int ROWS;
    int COLS;
};